File: mobile/android/app/src/main/java/com/tolkflip/models/Message.java
package com.tolkflip.models;

import androidx.room.Entity;
import androidx.room.PrimaryKey;
import androidx.room.TypeConverters;

import com.tolkflip.utils.Converters;

import java.util.List;
import java.util.Map;

@Entity(tableName = "messages")
@TypeConverters(Converters.class)
public class Message {
    @PrimaryKey(autoGenerate = false)
    private String messageId;
    private String threadId;
    private String senderId;
    private String receiverId;
    private String content;
    private String translatedContent;
    private String originalLanguage;
    private String targetLanguage;
    private MessageType messageType;
    private MessageStatus status;
    private long timestamp;
    private boolean isEncrypted;
    private byte[] encryptedContent;
    private List<String> mediaUrls;
    private Map<String, String> metadata;
    private String emotion;
    private boolean showOriginal;

    public enum MessageType {
        TEXT, IMAGE, VIDEO, AUDIO, FILE, VOICE_NOTE
    }

    public enum MessageStatus {
        SENDING, SENT, DELIVERED, READ, FAILED
    }

    public Message() {
        this.timestamp = System.currentTimeMillis();
        this.status = MessageStatus.SENDING;
        this.showOriginal = false;
    }

    // Getters and Setters
    public String getMessageId() {
        return messageId;
    }

    public void setMessageId(String messageId) {
        this.messageId = messageId;
    }

    public String getThreadId() {
        return threadId;
    }

    public void setThreadId(String threadId) {
        this.threadId = threadId;
    }

    public String getSenderId() {
        return senderId;
    }

    public void setSenderId(String senderId) {
        this.senderId = senderId;
    }

    public String getReceiverId() {
        return receiverId;
    }

    public void setReceiverId(String receiverId) {
        this.receiverId = receiverId;
    }

    public String getContent() {
        return content;
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getTranslatedContent() {
        return translatedContent;
    }

    public void setTranslatedContent(String translatedContent) {
        this.translatedContent = translatedContent;
    }

    public String getOriginalLanguage() {
        return originalLanguage;
    }

    public void setOriginalLanguage(String originalLanguage) {
        this.originalLanguage = originalLanguage;
    }

    public String getTargetLanguage() {
        return targetLanguage;
    }

    public void setTargetLanguage(String targetLanguage) {
        this.targetLanguage = targetLanguage;
    }

    public MessageType getMessageType() {
        return messageType;
    }

    public void setMessageType(MessageType messageType) {
        this.messageType = messageType;
    }

    public MessageStatus getStatus() {
        return status;
    }

    public void setStatus(MessageStatus status) {
        this.status = status;
    }

    public long getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(long timestamp) {
        this.timestamp = timestamp;
    }

    public boolean isEncrypted() {
        return isEncrypted;
    }

    public void setEncrypted(boolean encrypted) {
        isEncrypted = encrypted;
    }

    public byte[] getEncryptedContent() {
        return encryptedContent;
    }

    public void setEncryptedContent(byte[] encryptedContent) {
        this.encryptedContent = encryptedContent;
    }

    public List<String> getMediaUrls() {
        return mediaUrls;
    }

    public void setMediaUrls(List<String> mediaUrls) {
        this.mediaUrls = mediaUrls;
    }

    public Map<String, String> getMetadata() {
        return metadata;
    }

    public void setMetadata(Map<String, String> metadata) {
        this.metadata = metadata;
    }

    public String getEmotion() {
        return emotion;
    }

    public void setEmotion(String emotion) {
        this.emotion = emotion;
    }

    public boolean isShowOriginal() {
        return showOriginal;
    }

    public void setShowOriginal(boolean showOriginal) {
        this.showOriginal = showOriginal;
    }

    public String getDisplayContent() {
        if (showOriginal || translatedContent == null || translatedContent.isEmpty()) {
            return content;
        }
        return translatedContent;
    }
}

File: mobile/ios/Tolkflip/Models/Message.swift
import Foundation
import RealmSwift

enum MessageType: String, PersistableEnum, Codable {
    case text
    case image
    case video
    case audio
    case file
    case voiceNote
}

enum MessageStatus: String, PersistableEnum, Codable {
    case sending
    case sent
    case delivered
    case read
    case failed
}

class Message: Object, Codable {
    @Persisted(primaryKey: true) var messageId: String = UUID().uuidString
    @Persisted var threadId: String = ""
    @Persisted var senderId: String = ""
    @Persisted var receiverId: String = ""
    @Persisted var content: String = ""
    @Persisted var translatedContent: String?
    @Persisted var originalLanguage: String = "en"
    @Persisted var targetLanguage: String?
    @Persisted var messageType: MessageType = .text
    @Persisted var status: MessageStatus = .sending
    @Persisted var timestamp: Date = Date()
    @Persisted var isEncrypted: Bool = false
    @Persisted var encryptedContent: Data?
    @Persisted var mediaUrls: List<String>
    @Persisted var emotion: String?
    @Persisted var showOriginal: Bool = false

    var displayContent: String {
        if showOriginal || translatedContent == nil || translatedContent!.isEmpty {
            return content
        }
        return translatedContent!
    }

    enum CodingKeys: String, CodingKey {
        case messageId, threadId, senderId, receiverId, content
        case translatedContent, originalLanguage, targetLanguage
        case messageType, status, timestamp, isEncrypted
        case encryptedContent, emotion, showOriginal
    }
}

class ChatThread: Object, Codable {
    @Persisted(primaryKey: true) var threadId: String = UUID().uuidString
    @Persisted var otherUserId: String = ""
    @Persisted var otherUserName: String = ""
    @Persisted var otherUserAvatar: String = ""
    @Persisted var lastMessage: String = ""
    @Persisted var lastMessageTime: Date = Date()
    @Persisted var unreadCount: Int = 0
    @Persisted var preferredLanguage: String?
    @Persisted var showOriginal: Bool = false
    @Persisted var enableEmotionDetection: Bool = true
    @Persisted var isArchived: Bool = false
}
